google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/SingleUseTransaction.java

```diff
@@ -424,32 +422,22 @@ private TransactionalBatchUpdateCallable(Iterable<ParsedStatement> updates) {
     // handle timeouts and canceling of a statement.
     while (true) {
       try {
+        return asyncExecuteStatement(update, callable);
+      } catch (AbortedException e) {
         try {
-          return asyncExecuteStatement(update, callable);
-        } catch (AbortedException e) {
-          try {
-            Thread.sleep(e.getRetryDelayInMillis() / 1000);
-          } catch (InterruptedException e1) {
-            throw SpannerExceptionFactory.newSpannerException(
-                ErrorCode.CANCELLED, "Statement execution was interrupted", e1);
-          }
-          // Check whether the timeout time has been exceeded.
-          long executionTime = System.currentTimeMillis() - startedTime;
-          if (getStatementTimeout().hasTimeout()
-              && executionTime > getStatementTimeout().getTimeoutValue(TimeUnit.MILLISECONDS)) {
-            throw SpannerExceptionFactory.newSpannerException(
-                ErrorCode.DEADLINE_EXCEEDED,
-                "Statement execution timeout occurred for " + update.getSqlWithoutComments());
-          }
+          Thread.sleep(e.getRetryDelayInMillis() / 1000);
+        } catch (InterruptedException e1) {
+          throw SpannerExceptionFactory.newSpannerException(
+              ErrorCode.CANCELLED, "Statement execution was interrupted", e1);
         }
-      } catch (Exception e) {
-        if (txManager != null) {
-          if (txManager.getState()
-              == com.google.cloud.spanner.TransactionManager.TransactionState.STARTED) {
-            fireAndForgetRollbackAndCloseTxManager(txManager);
```

Comment:
Can you comment on what the performance implication for the client is if the rollback is not triggered immediately? We may be locking the rows for longer than is necessary while we wait for the backend to abort the transaction which might prevent the user from doing a manual retry.

Reply:
The rollback is triggered immediately, but I understand why that is not obvious from the code. The `TransactionManager.close()` method that is called on line 381 will execute a rollback if the transaction is still active. I've added a comment there to make that clearer.